### DISCUSSION TITLE
Different type of return with same configuration RTU over TCP

### DISCUSSION THREAD

**Discussion by rosariok24** [aldas/modbus-tcp-client#98] [Q&A]
Hello Aldas,

i'm in conversion from a Client in Modbus TCP/ip with a Client in Modbus RTU over TCP. 
Until i used another type of class " Modbus Master" whit Modbus TCP in FC4, i have all the register from the PLC in the correct way.
Now i'm attached with a converter that i have the only possibility with RTU over TCP.

Below there is che script modded to connect to the register "0" and turn back only 1 byte, just to realize if the result is what i expect.

 
```php
use ModbusTcpClient\Network\BinaryStreamConnection;
use ModbusTcpClient\Packet\ModbusFunction\ReadInputRegistersRequest;
use ModbusTcpClient\Packet\ModbusFunction\ReadInputRegistersResponse;
use ModbusTcpClient\Packet\ResponseFactory;
use ModbusTcpClient\Packet\RtuConverter;

require __DIR__ . '/../vendor/autoload.php';

$connection = BinaryStreamConnection::getBuilder()
    ->setPort(7109)
    ->setHost('192.168.1.51')
    ->setReadTimeoutSec(3) // increase read timeout to 3 seconds
    ->setIsCompleteCallback(function ($binaryData, $streamIndex) {
        return true;
    })
    ->build();

$startAddress = 0;
$quantity = 1;
$slaveId = 1;

$rtuPacket = RtuConverter::toRtu(new ReadInputRegistersRequest($startAddress, $quantity, $slaveId));
$binaryData = $connection->connect()->sendAndReceive($rtuPacket);
$responseAsTcpPacket = RtuConverter::fromRtu($binaryData);

try {
    $binaryData = $connection->connect()->sendAndReceive($rtuPacket);
    echo 'RTU Binary received (in hex):   ' . unpack('H*', $binaryData)[1] . PHP_EOL;
    $response = RtuConverter::fromRtu($binaryData);
    echo 'Parsed packet (in hex):     ' . $response->toHex() . PHP_EOL;
    echo 'Data parsed from packet (bytes):' . PHP_EOL;
    print_r($response->getData());

} catch (Exception $exception) {
    echo 'An exception occurred' . PHP_EOL;
    echo $exception->getMessage() . PHP_EOL;
    echo $exception->getTraceAsString() . PHP_EOL;
} finally {
    $connection->close();
}
```

At the execution i have some trouble, because i want a byte and in the print i have 2 element in out

![result_rtu call](https://user-images.githubusercontent.com/94061480/141152381-5a7208a6-dea2-4c48-a6c3-8004b538e4a4.png)

the second element i expect was more greater like thousands not hundreds. In the image below there is the same call with modbus pool at the same register in FC4 to have the correct data in out.

![modbus-pool result](https://user-images.githubusercontent.com/94061480/141152605-e70a602d-856c-4afa-8a55-49d8c5872562.png)
![modbus-pool readwrite conf](https://user-images.githubusercontent.com/94061480/141152628-152f7694-e39a-4c21-81e2-fe3e53ec20aa.png)
![modbus-pool connection conf](https://user-images.githubusercontent.com/94061480/141152651-bcd93c9b-f16f-4c41-9055-34279b0cc41d.png)

There is something that is missing for my side.

**Comment by aldas** [comment]
When you use `print_r($response->getData());` it will print each byte in response. In modbus 1 register (or word) is 16bit memory range, that means 2 bytes combined is 1 modbus register. You need to use methods from API to convert raw bytes to your needed data types.


See this part how raw bytes can be transformed to different data types:
https://github.com/aldas/modbus-tcp-client/blob/c39a0e51519c5ef76f6ff0f2fb8e36fb34de2935/examples/index.php#L74
